### PR TITLE
[BEAM-3371] Enable running integration tests on Spark

### PR DESCRIPTION
**Pull request by lgajowy**
This PR enables running IOIT on Spark runner. One can do that by setting up remote spark cluster (spark master) and running this command: 

```
./gradlew clean integrationTest -p sdks/java/io/file-based-io-tests/ -DintegrationTestPipelineOptions='["--numberOfRecords=1000", "--filenamePrefix=PREFIX", "--runner=TestSparkRunner", "--sparkMaster=spark://LGs-Mac.local:7077", "--tempLocation=/tmp/"]' -DintegrationTestRunner=spark --tests org.apache.beam.sdk.io.text.TextIOIT --info
```

I experienced some difficulties with TFRecordIOIT, ParquetIOIT and XmlIOIT tests: 
 - XmlIOIT fails on assertion (hashcode of PCollection is different that it should be)
 - ParquetIOIT suffers for dependency missmatch (`java.lang.NoSuchMethodError: org.apache.parquet.hadoop.ParquetWriter$Builder.<init>(Lorg/apache/parquet/io/OutputFile;)V`)
 - TFRecordIOIT cannot find the created file: `java.io.FileNotFoundException: No files found for spec: PREFIX_1534520885787*`

Those issues are of different nature than the one described in BEAM-3371 and have to be tackled separately. 

@iemejia Could you take a look when you're back? There's no need to rush - I'll be off for 2 weeks now. 

CC: @pabloem 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by lgajowy** [comment]
BTW: what is the reason of the fact that filesToStage option is duplicated in `DataflowWorkerPoolOptions`, `FlinkPipelineOptions` and `SparkPipelineOptions`? Do you think we can move this to `PipelineOptions` interface? IMO it would make things even easier (maybe not only for this fix).

**Comment by lgajowy** [comment]
Run Java PreCommit

**Comment by lgajowy** [comment]
Created issues for failing tests: 
https://issues.apache.org/jira/browse/BEAM-5165
https://issues.apache.org/jira/browse/BEAM-5164
https://issues.apache.org/jira/browse/BEAM-5163

**Comment by lgajowy** [comment]
@iemejia rebased and fixed the spotless issue. Could you take a look?

**Comment by lgajowy** [comment]
@iemejia ping :)

Alternatively, (if you're very busy), please suggest some other reviewer.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironment.java [hunk 1]
```diff
@@ -95,73 +83,33 @@ public void translate(Pipeline pipeline) {
...
-      ZipFiles.zipDirectory(directoryToStage, new FileOutputStream(uniqueDirectoryPath));
-    } catch (IOException e) {
-      throw new RuntimeException(e);
+  /**
+   * Local configurations work in the same JVM and have no problems with improperly formatted files
+   * on classpath (eg. directories with .class files or empty directories). Prepare files for
+   * staging only when using remote cluster.
+   */
+  private void prepareFilesToStageForRemoteClusterExecution() {
+    if (!options.getFlinkMaster().matches("\\[.*\\]")) {
```
Comment (aromanenko-dev) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironment.java]: Can we just check that `getFlinkMaster()` is not empty?

Comment (lgajowy) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironment.java]: So the idea here is that we don't want to `prepareFilesForStaging()` when using local (and auto, which in turn can decide to run [local]). We only need to prepare resources like this (remove nonexistent paths, package test resources into jars etc) when we actually stage them to a remote runner. 

I figured that it might be more descriptive when the code will be more explicit and mention "[auto]" and "[local]" here, wdyt?

**Comment by lgajowy** [comment]
Hi @aromanenko-dev - I applied the suggestions. PTAL.

**Comment by aromanenko-dev** [comment]
Retest this please

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironment.java (hunk 1 above)
Comment (aromanenko-dev) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironment.java]: I agree with `[local]` but not sure about `[auto]` since it depends on the environment and, in the end, it could be run on remote cluster. In this case, staging won't be applied, right?

Comment (lgajowy) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironment.java]: This is unfortunately true. I must admit that this is a kind of a bitter-sweet compromise to me but couldn't find some other sensible option. 

I didn't find a better way to recognize if we're running on a local or remote cluster. I don't think that this deserves eg. a separate pipeline option (`eg --integrationTest`) due to two reasons: 
 - this is a production runner, not a test one so I wouldn't add such test-related flags to it because it is a "noise" for runner users
 - we are not sure if this mechanism won't be useful in some other cases

I decided to do it like this at least for now, leaving a descriptive comment for the future. We can always develop it more. Do you think this is enough?

BTW: I was missing [collection] here - it also shouldn't be used. I'll post the fix.

**Comment by lgajowy** [comment]
@aromanenko-dev I amended it a little bit and left a comment (ptal).

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironment.java (hunk 1 above)
Comment (aromanenko-dev) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironment.java]: I'm ok with that but I'd double check if this doesn't break something. @mxm, wdyt?

**Comment by lgajowy** [comment]
@mxm @aromanenko-dev thanks! 

I added tests for Flink. Due to the fact that Spark runner invokes the code in run() method and creates SparkContext there, I had some trouble with creating similar tests for Spark too so i didn't submit them in this PR. Maybe we can do this later. I got the following error when I run similar tests: 

> Cannot reuse spark context with different spark master URL. Existing: local[1], requested: spark://localhost:7077. 

I tried to reuse the context (with `ReuseContextRule.java` class). I also tried to stop the context somehow but all this lead me nowhere. The code I used is here: https://github.com/apache/beam/compare/master...lgajowy:spark-integration-tests-2?expand=1#diff-9336eb87a4aea9ba0f254a1318f1fc90 

@mxm @aromanenko-dev could you take a look again?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java [hunk 2]
```diff
@@ -435,7 +451,7 @@ public void visitPrimitiveTransform(TransformHierarchy.Node node) {
     protected <TransformT extends PTransform<? super PInput, POutput>>
         TransformEvaluator<TransformT> translate(
             TransformHierarchy.Node node, TransformT transform, Class<TransformT> transformClass) {
-      //--- determine if node is bounded/unbounded.
+      // --- determine if node is bounded/unbounded.
```
Comment (mxm) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java]: unnecessary change

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironmentTest.java [hunk 3]
```diff
@@ -49,6 +63,7 @@ public void shouldRecognizeAndTranslateStreamingPipeline() {
         .apply(
             ParDo.of(
                 new DoFn<Long, String>() {
+
```
Comment (mxm) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironmentTest.java]: unnecessary change

File: runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java (hunk 2 above)
Comment (lgajowy) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java]: reverted.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironmentTest.java (hunk 3 above)
Comment (lgajowy) [reply runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironmentTest.java]: reverted.

File: runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironmentTest.java [hunk 4]
```diff
@@ -61,4 +76,63 @@ public void processElement(ProcessContext c) throws Exception {
...
+  public void shouldNotPrepareFilesToStagewhenFlinkMasterIsSetToCollection() throws IOException {
+    FlinkPipelineOptions options = testPreparingResourcesToStage("[collection]");
+
+    assertThat(options.getFilesToStage().size(), is(2));
+    assertThat(options.getFilesToStage(), everyItem(not(matches(".*\\.jar"))));
+  }
+
+  @Test
+  public void shouldNotPrepareFilesToStageWhenFlinkMasterIsSetToLocal() throws IOException {
+    FlinkPipelineOptions options = testPreparingResourcesToStage("[local]");
```
Comment (aromanenko-dev) [inline runners/flink/src/test/java/org/apache/beam/runners/flink/FlinkPipelineExecutionEnvironmentTest.java]: These 3 similar tests above could be only one parameterised but not a big deal for now.

**Comment by aromanenko-dev** [comment]
Since Flink side is ok (thanks to @mxm for review) and other part is fine for me as well, so it LGTM and let's wait for green tests and I think we can merge it after.

**Comment by lgajowy** [comment]
Run Java PreCommit

**Comment by lgajowy** [comment]
@aromanenko-dev I  rewrote the history squashing the commits that were irrelevant. Initially, this PR was all about Spark (BEAM-3371) but I also added one commit related to Flink (BEAM-3370) which is about adding the tests. All this in light of the recent [dev list discussion](https://lists.apache.org/thread.html/6d922820d6fc352479f88e5c8737f2c8893ddb706a1e578b50d28948@%3Cdev.beam.apache.org%3E) and [docs](https://beam.apache.org/contribute/committer-guide/#finishing-touches).

Let me know if this looks good to you too - I have never really received feedback about the history shape, although I try to care about it. It looks that we (commiters especially) should be more aware about the git history so this PR is a good chance to actually get the feedback and practice this. :)

**Comment by aaltay** [comment]
What is the status of this PR?

**Comment by lgajowy** [comment]
@aaltay @aromanenko-dev suggested some additional manual checks what happens if we run this using spark-submit. I didn't test that yet - I should be able to do it later this week.

**Comment by aromanenko-dev** [comment]
Run Flink ValidatesRunner

**Comment by aromanenko-dev** [comment]
@lgajowy I performed testing on my side - I created a fat jar with basic WordBasic pipeline using mvn artifacts built from the branch of this PR and run it on my virtual YARN/Spark cluster. I didn't see any issues with that, so, I think it LGTM and we can merge it.

**Comment by lgajowy** [comment]
@aromanenko-dev thanks a lot! :)
